langchain/llms/beam.py

```diff
@@ -0,0 +1,299 @@
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of llm."""
+        return "Beam"
+
+    def app_creation(self) -> None:
+        """Creates a Python file which will contain your Beam app definition."""
+        script = textwrap.dedent(
+            """\
+        import beam
+
+        # The environment your code will run on
+        app = beam.App(
+            name="{name}",
+            cpu={cpu},
+            memory="{memory}",
+            gpu="{gpu}",
+            python_version="{python_version}",
+            python_packages={python_packages},
+        )
+
+        app.Trigger.RestAPI(
+            inputs={{"prompt": beam.Types.String(), "max_length": beam.Types.String()}},
+            outputs={{"text": beam.Types.String()}},
+            handler="run.py:beam_langchain",
+        )
+
+        """
+        )
+
+        script_name = "app.py"
+        with open(script_name, "w") as file:
+            file.write(
+                script.format(
+                    name=self.name,
+                    cpu=self.cpu,
+                    memory=self.memory,
+                    gpu=self.gpu,
+                    python_version=self.python_version,
+                    python_packages=self.python_packages,
+                )
+            )
+
+    def run_creation(self) -> None:
+        """Creates a Python file which will be deployed on beam."""
+        script = textwrap.dedent(
+            """
+        import os
+        import transformers
+        from transformers import GPT2LMHeadModel, GPT2Tokenizer
+
+        model_name = "{model_name}"
+
+        def beam_langchain(**inputs):
+            prompt = inputs["prompt"]
+            length = inputs["max_length"]
+
+            tokenizer = GPT2Tokenizer.from_pretrained(model_name)
+            model = GPT2LMHeadModel.from_pretrained(model_name)
+            encodedPrompt = tokenizer.encode(prompt, return_tensors='pt')
+            outputs = model.generate(encodedPrompt, max_length=int(length),
+              do_sample=True, pad_token_id=tokenizer.eos_token_id)
+            output = tokenizer.decode(outputs[0], skip_special_tokens=True)
+
+            print(output)
+            return {{"text": output}}
+
+        """
+        )
+
+        script_name = "run.py"
+        with open(script_name, "w") as file:
+            file.write(script.format(model_name=self.model_name))
+
+    def _deploy(self) -> str:
+        """Call to Beam."""
+        try:
+            import beam  # type: ignore
+
+            if (beam.__path__) == "":
+                raise ImportError
+        except ImportError:
+            raise ValueError(
+                "Could not import beam python package. "
+                "Please install it with `curl "
+                "https://raw.githubusercontent.com/slai-labs"
+                "/get-beam/main/get-beam.sh -sSfL | sh`."
+            )
+
+        self.app_creation()
+        self.run_creation()
+
+        process = subprocess.run(
```

Comment:
is there any way to avoid writing scripts and kicking of subprocesses? for example could we make the script in run_creation a standalone function and add to documentation that users should run that before instantiating the LLM?

Reply:
I'm not sure that this is the best way to approach it—a standalone static script wouldn't have the remote machine specs passed in during the instantiation of the LLM and would need to be updated anyways. Since the _deploy method isn't a required call(you can pass in already deployed app_ids to the _call method) would leaving this be alright?

Reply:
aren't most of the params used in script creation only on the LLM for script creation (like they're not used in the actual calls, right?). so you could have those params as args on the standalone function and remove them from the LLM?

happy to defer to you on what's best though

Reply:
Structurally, where would you want the standalone app_creation and run_creation functions to be? I can see how that might be preferred, just want to make sure that we structure it intuitively.

Reply:
im ok keeping it in this llms/beam.py file or adding a new utilities/beam.py file for them